lib/ssh/run_parse.go

```diff
@@ -0,0 +1,139 @@
+package sshutils
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"golang.org/x/crypto/ssh"
+)
+
+type OutputParseFn func(r *bufio.Reader) (out interface{}, err error)
+type LogFnType func(format string, args ...interface{})
+
+const exitStatusUndefined = -1
+
+// Run is a simple method to run external program and don't care about its output or exit status
+func Run(ctx context.Context, logFn LogFnType, client *ssh.Client, cmd string, env map[string]string) error {
+	_, exit, err := RunAndParse(ctx, logFn, client,
+		cmd, env, ParseDiscard)
+
+	if err != nil {
+		return trace.Wrap(err, cmd)
+	}
+
+	if exit != 0 {
+		return trace.Errorf("%s returned %d", cmd, exit)
+	}
+
+	return nil
+}
+
+// RunAndParse runs remote SSH command with environment variables set by `env`
+// exitStatus is -1 if undefined
+func RunAndParse(ctx context.Context, logFn LogFnType, client *ssh.Client, cmd string, env map[string]string, parse OutputParseFn) (out interface{}, exitStatus int, err error) {
```

Comment:
The function expects a `ssh.Session`, and the logging detail can be setup prior to the call:
```go
session, err := client.NewSession()
if err != nil {
	return trace.Wrap(err, "failed to create a session")
}
entry := log.WithFields(log.Fields{"client": client.RemoteAddr()})
exitStatus, err := RunAndParse(ctx, session, cmd, env, parseFn, entry)

func RunAndParse(ctx context.Context, session *ssh.Session, cmd string, env map[string]string, parse OutputParseFn, entry *log.Entry) (out interface{}, exitStatus int, err error) {
	// ...
	entry.Debugf("stderr: %s", out)
}
```
This way you don't need to spill too much detail into the function itself.

And the function itself feels too heavy.